docs/blogs/elasticdl-usability.md

```diff
@@ -0,0 +1,309 @@
+ --distribution_strategy=ParameterServerStrategy \
+ --output=model_output
+ ```
+
+ 在上述例子中，我们指定了 Parameter Server 的分布式策略，由一个parameter server 和 两个 worker 共同完成训练任务。
+ ElasticDL 的 master pod 将会被首先创建，然后由 master 负责启动 worker pod，以及 parameter server pod，并且建立通信。
+ ElasticDL master 可以监控每个 pod 的状态，当有 pod 挂掉时，master 会重新拉起新的 pod。
+
+## Parameter Server 的改进
+
+在搜索广告等场景，模型中可能包含较大的 embedding table，其内存会超过单机内存。我们通常使用 Parameter Server (PS) 分布式策略来训练此类模型。
+在 PS 策略下，PS 上存储着模型参数，worker 从 PS 上请求参数。
+worker 在本地使用训练数据计算梯度之后，把梯度再发送到 PS 上，PS 使用 worker 传来的梯度来迭代更新模型参数。
+
+ElasticDL 用 Go 实现了 Parameter Server，具有良好的吞吐能力和可扩展性。并且，我们针对 embedding table 做了一些额外的优化。
+
+- embedding vector 惰性初始化，用户无需提前指定 embedding table 的大小
+- 把一个 embedding table 拆分到多个 PS 上存储与更新，均衡存储与通信的负载
+- worker 从 PS 请求参数时，先滤除重复 ID ，只请求不同的参数，减少通信量
+- worker 向 PS 发送梯度时，本地先把相同 ID 的梯度进行合并，减少通信量
+
+通过上述设计与实现，ElasticDL 可以很高效的完成搜索推荐广告模型的训练。
+我们用一个推荐中常用的 deepFM 模型为例，来说明 ElasticDL 相比于去年9月开源时的性能提升。
+
+**TODO** 添加性能对比
+
+## 使用 ElasticDL 进行 Kaggle 实战
+
+在本小节中，我们将使用 ElasticDL 进行一次 Kaggle 实战。
+本例中使用的是 Kaggle 上 Display Advertising Challenge 中的 criteo 数据集，这是一个关于广告点击率预估的比赛。
+我们将使用 xDeepFM 模型来进行建模，所有的实例代码都放在了 ElasticDL 的 [model zoo](https://github.com/sql-machine-learning/elasticdl/tree/develop/model_zoo/dac_ctr)中。
+
+### 数据预处理
+
+1. 我们首先从官方
+[链接](https://labs.criteo.com/2014/02/download-kaggle-display-advertising-challenge-dataset)
+下载 criteo 数据集。
+
+1. 然后我们需要把原始数据转换为 RecordIO 文件格式。
+我们提供了如下的转换脚本：
+
+```bash
+python convert_to_recordio.py \
+    --records_per_shard 400000 \
+    --output_dir ./dac_records \
+    --data_path train.txt
+```
+
+原始数据会被按照 19:1 的比例，拆分为训练集和验证集，
+转换后的数据放在 dac_records 目录中。
+
+1. 对原始数据进行特征统计。对于连续的特征，我们统计得出均值和方差；
+对于离散的特征，我们得出特征值个数。我们把统计后的数据放在一个文件中，供后续使用。
+
+### 模型定义
+
+xDeepFM 模型由三部分组成，分别是 linear logits，dnn logits 和 xfm logits。
+借助 Keras API，我们可以很清晰的描述模型结构。
+这里贴出 dnn logits 部分的描述代码，完整的模型定义可以参见 model zoo。
+
+```python
+deep_embeddings = lookup_embedding_func(
+    id_tensors, max_ids, embedding_dim=deep_embedding_dim,
+)
+dnn_input = tf.reshape(
+    deep_embeddings, shape=(-1, len(deep_embeddings) * deep_embedding_dim)
+)
+if dense_tensor is not None:
+    dnn_input = tf.keras.layers.Concatenate()([dense_tensor, dnn_input])
+
+dnn_output = DNN(hidden_units=[16, 4], activation="relu")(dnn_input)
+
+dnn_logit = tf.keras.layers.Dense(1, use_bias=False, activation=None)(
+    dnn_output
+)
+```
+
+### 提交训练任务
+
+我们首先在 Google Cloud 上创建一个 GKE 集群，并且把转换好的 RecordIO 训练数据上传到集群上。
+详细的过程可以参考 ElasticDL 的 [gcloud教程](https://github.com/sql-machine-learning/elasticdl/blob/develop/docs/tutorials/elasticdl_cloud.md)。
+
+然后，我们在本地制作一个镜像，该镜像包含了 xDeepFM 模型定义，以及相关依赖包。
+
+```bash
+FROM tensorflow
+RUN pip install elasticdl
+COPY model_zoo /model_zoo
+```
+
+我们需要把该镜像推送到 GKE 集群能够访问到的仓库中，比如说 docker hub 的仓库中。
+
+最后，我们通过 ElasticDL client 工具向 GKE 集群提交训练作业。
+我们使用 ParameterServer 分布式策略进行训练，有 2 个 parameter serve pods 和 5个 worker pods共同参与训练。
+
+```bash
+elasticdl train \
+  --image_name=${your_docker_hub_id}/elasticdl:ci \
```

Comment:
This should be docker hub registry instead of docker hub id?

Reply:
Done